### PR TITLE
[SPARK-27162][SQL] Add new method asCaseSensitiveMap in CaseInsensitiveStringMap

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Currently, DataFrameReader/DataFrameReader supports setting Hadoop configurations via method `.option()`. 
E.g, the following test case should be passed in both ORC V1 and V2
```
  class TestFileFilter extends PathFilter {
    override def accept(path: Path): Boolean = path.getParent.getName != "p=2"
  }

  withTempPath { dir =>
      val path = dir.getCanonicalPath

      val df = spark.range(2)
      df.write.orc(path + "/p=1")
      df.write.orc(path + "/p=2")
      val extraOptions = Map(
        "mapred.input.pathFilter.class" -> classOf[TestFileFilter].getName,
        "mapreduce.input.pathFilter.class" -> classOf[TestFileFilter].getName
      )
      assert(spark.read.options(extraOptions).orc(path).count() === 2)
    }
  }
```
While Hadoop Configurations are case sensitive, the current data source V2 APIs are using `CaseInsensitiveStringMap` in the top level entry `TableProvider`. 
To create Hadoop configurations correctly, I suggest 
1. adding a new method `asCaseSensitiveMap` in `CaseInsensitiveStringMap`.
2. Make `CaseInsensitiveStringMap` read-only to ambiguous conversion in `asCaseSensitiveMap`

## How was this patch tested?

Unit test

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 1]
```diff
@@ -78,11 +81,13 @@ public String get(Object key) {
 
   @Override
   public String put(String key, String value) {
+    originalMap.put(key, value);
```
Comment (gengliangwang) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: Things become a bit ambiguous here. E.g. after `put("a", "b")` and `put("A", "B")` to a empty `CaseInsensitiveStringMap`, `originalMap` has two new entries, while  `delegate` has only one new entry.
To me, this is the simplest way to maintain `originalMap`. Suggestions are welcomed.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 2]
```diff
@@ -78,11 +81,13 @@ public String get(Object key) {
...
     return delegate.put(toLowerCase(key), value);
   }
 
   @Override
   public String remove(Object key) {
+    originalMap.remove(key);
```
Comment (gengliangwang) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: Things become a bit ambiguous here too.

**Comment by gengliangwang** [comment]
@rdblue @cloud-fan

**Comment by rdblue** [comment]
@gengliangwang, is it required that Hadoop configuration options are passed through this map or is there another way?

**Comment by gengliangwang** [comment]
Yes, I think so.
`DataFrameReader.load()` => `TableProvider.getTable()` => ...

**Comment by cloud-fan** [comment]
AFAIK hadoop conf can be set in 3 ways:
1. global level, via `SparkContext.hadoopConfiguration`
2. session level, via `SparkSession.conf`
3. operation level, via `DataFrameReader/Writer.option`

1 and 2 are fine, as they are case sensitive. The problem is 3, as data source v2 treats options as case-insensitive.

There are 2 solutions I can think of
1. Do not support operation level hadoop conf for data source v2.
2. Keep the original case sensitive map.

I think 2 is more reasonable, which is this PR trying to do.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 3]
```diff
@@ -40,9 +40,12 @@ public static CaseInsensitiveStringMap empty() {
     return new CaseInsensitiveStringMap(new HashMap<>(0));
   }
 
+  private final Map<String, String> original;
+
   private final Map<String, String> delegate;
 
   public CaseInsensitiveStringMap(Map<String, String> originalMap) {
+    this.original = new HashMap<>(originalMap);
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: this should be `new HashMap<>(originalMap.size);`, otherwise we add data to it twice.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala [hunk 4]
```diff
@@ -96,6 +98,11 @@ private[sql] class SessionState(
     hadoopConf
   }
 
+  def newHadoopConfWithCaseInsensitiveOptions(options: CaseInsensitiveStringMap): Configuration = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: I don't think we should pollute `SessionState` with the case insensitive map stuff. Can we inline this method?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: I don't think we should pollute `SessionState` with the case insensitive map stuff. Can we inline this method?

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: Otherwise, developers might not be aware of using `.getOriginalMap` if they want to create Hadoop configuration from CaseInsensitiveStringMap.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 5]
```diff
@@ -78,11 +81,13 @@ public String get(Object key) {
 
   @Override
   public String put(String key, String value) {
+    original.put(key, value);
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: The thing worries me most is the inconsistency between the case insensitive map and the original map. I think we should either fail or keep the latter entry if `a -> 1, A -> 2` appears together.

One thing we can simplify is, `CaseInsensitiveStringMap` is read by data source and can be read-only. Then it can be easier to resolve conflicting entries at the beginning.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: Then we should document it in `CaseInsensitiveMap`. data source developers can't access `SessionState`

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java (hunk 5 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: But the method `put/remove/clear` still need to be implemented...Do you mean that we can ignore original map in these method ?

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: We can just throw exception in these methods, and say this map is readonly.

Comment (gengliangwang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: @cloud-fan This is a good solution! 
@rdblue What do you think?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala (hunk 4 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: I agree with @cloud-fan, I'd rather use `newHadoopConfWithOptions` and not add a new method here.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 6]
```diff
@@ -157,4 +163,11 @@ public double getDouble(String key, double defaultValue) {
     String value = get(key);
     return value == null ? defaultValue : Double.parseDouble(value);
   }
+
+  /**
+   * Returns the original case-sensitive map.
+   */
+  public Map<String, String> getOriginalMap() {
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: What about `asCaseSensitiveMap`? That is more clear that using "original" because that doesn't indicate why someone would call this method.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java (hunk 5 above)
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: That works for me. It is always a good idea to avoid passing mutable state to plugin code.

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 7]
```diff
@@ -40,11 +40,16 @@ public static CaseInsensitiveStringMap empty() {
...
+  private final Map<String, String> original;
+
   private final Map<String, String> delegate;
 
   public CaseInsensitiveStringMap(Map<String, String> originalMap) {
-    this.delegate = new HashMap<>(originalMap.size());
-    putAll(originalMap);
+    original = new HashMap<>(originalMap);
+    delegate = new HashMap<>(originalMap.size());
+    for (Map.Entry<? extends String, ? extends String> entry : originalMap.entrySet()) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: is the `? extends String` required? Can we just use `String`?

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 8]
```diff
@@ -40,11 +40,16 @@ public static CaseInsensitiveStringMap empty() {
...
+      delegate.put(toLowerCase(entry.getKey()), entry.getValue());
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: If the key is already in the case insensitive map, we should fail and say duplicated keys detected.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 9]
```diff
@@ -78,24 +83,22 @@ public String get(Object key) {
 
   @Override
   public String put(String key, String value) {
-    return delegate.put(toLowerCase(key), value);
+    throw new UnsupportedOperationException();
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: let's add the error message to say the map is read only.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala [hunk 10]
```diff
@@ -32,7 +34,7 @@ import org.apache.spark.sql.catalyst.parser.ParserInterface
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.execution._
 import org.apache.spark.sql.streaming.StreamingQueryManager
-import org.apache.spark.sql.util.{ExecutionListenerManager, QueryExecutionListener}
+import org.apache.spark.sql.util.{CaseInsensitiveStringMap, ExecutionListenerManager, QueryExecutionListener}
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala]: unnecessary changes

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java (hunk 9 above)
Comment (gengliangwang) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: This is following `UnmodifiableMap`.

Comment (gengliangwang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: How about showing warning message here?

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: Why do we need a warning message right before throwing an exception?

Comment (gengliangwang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: E.g.
```
spark.read.option("foo", "bar").option("fOO", "bar")....
```
This is valid.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java (hunk 8 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: How about showing warning message here?
E.g.
```
spark.read.option("foo", "bar").option("fOO", "bar")....
```
This is valid.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 11]
```diff
@@ -35,16 +37,29 @@
  */
 @Experimental
 public class CaseInsensitiveStringMap implements Map<String, String> {
+  private final Logger logger = LoggerFactory.getLogger(CaseInsensitiveStringMap.class);
+
+  private String unsupportedOperationMsg = "CaseInsensitiveStringMap is read-only.";
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: nit:
```
private void failAsReadonly() {
  throw new UnsupportedOperationException("CaseInsensitiveStringMap is read-only.")
}
```

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 12]
```diff
@@ -157,4 +170,11 @@ public double getDouble(String key, double defaultValue) {
     String value = get(key);
     return value == null ? defaultValue : Double.parseDouble(value);
   }
+
+  /**
+   * Returns the original case-sensitive map.
+   */
+  public Map<String, String> asCaseSensitiveMap() {
+    return original;
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: This should return a read-only version of original. You can use `Collections.unmodifiableMap`.

**Comment by rdblue** [comment]
Just one more problem: the case sensitive map should not allow modifications. Once that's fixed and tests are passing, +1.

File: sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java [hunk 13]
```diff
@@ -21,11 +21,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/util/CaseInsensitiveStringMap.java]: Minor: please don't use wildcard imports because it can lead to commit conflicts. It's also harder to see where symbols are coming from.

**Comment by rdblue** [comment]
+1 when tests are passing.

**Comment by cloud-fan** [comment]
thanks, merging to master!
